parser.js

```diff
@@ -184,9 +184,77 @@ class Extractor {
         return {'name': name, 'scope': scope, 'type': type, 'visibility': visibility}
     }
 
+    _isGenericType(string, index) {
+        let countLeftArrow = 0
+        let countRightArrow = 0
+
+        let stringGenericType = ''
+        for (let i = index; i < string.length; i++) {
+            stringGenericType += string[i]
+            if (string[i] === '<') {
+                countLeftArrow++
+            }
+            if (string[i] === '>') {
+                countRightArrow++
+            }
+            if (countLeftArrow === countRightArrow) {
+                for (let stringGenericTypeElement of stringGenericType) {
+                    if (stringGenericTypeElement === ',') {
+                        return {
+                            stringGenericType,
+                            index: i,
+                            isGenericType: 1
+                        }
+                    }
+                }
+                return {
+                    stringGenericType,
+                    index: i,
+                    isGenericType: 0
+                }
+            }
+        }
+    }
+
+    _parseCommaToBar(params) {
+        let refactorString = ''
+        let leftArrow = 0;
+
+        for (let i = 0; i < params.length; i++) {
+            if (!leftArrow) {
+                if (params[i] === ',') {
+                    refactorString += '|'
+                } else {
+                    if (params[i] !== '<') {
+                        refactorString += params[i]
+                    }
+                }
+            }
+            if (params[i] === '<') {
+                let objectGenericType = this._isGenericType(params, i)
+                if (objectGenericType.isGenericType) {
+                    refactorString += objectGenericType.stringGenericType
+                    i = objectGenericType.index + 1
+                }
+                leftArrow = 1
+            }
+            if (leftArrow) {
+                if (params[i] === ',') {
+                    refactorString += '|'
+                } else {
+                    if (params[i] !== undefined) {
+                        refactorString += params[i]
+                    }
+                }
+            }
```

Comment:
you can simplify this code:

```
        let refactorString = ''
        for (let i = 0; i < params.length; i++) {
            if (params[i] === '<') {
                let objectGenericType = this._isGenericType(params, i)
                if (objectGenericType.isGenericType) {
                    refactorString += objectGenericType.stringGenericType
                    i = objectGenericType.index + 1
                }
            }
            if (params[i] === ',') {
                refactorString += '|'
            } else {
                if (params[i] !== '<' && params[i] !== undefined) {
                    refactorString += params[i]
                }
            }
        }
        return refactorString
```

Also, we need to provide a way to show an error if the generics is malformed. Example:

- `localizaUsuarios(nome: str, teste2: Map<str, ArrayList<Str>): List<String>` - missing ending '>'
- `localizaUsuarios(nome: str, teste2: Map<str, ArrayList>Str>>): List<String>` - unbalanced or strange char (">")

Reply:
I don't know, without checkFirstArrow, it is not possible to have the exact precision of where to place and when to place or start the iterator with '<'. without it, some tests don't pass 

 ![](https://i.imgur.com/rgAyGuO.png) 

what about errors if the generics are badly formatted, don't you want to create another issue for that? I can do it right away.